### DISCUSSION TITLE
Feature: Provide extension API for custom usage scanner

### DISCUSSION THREAD

**Discussion by icanhazstring** [composer-unused/composer-unused#69] [Ideas]
## Is your feature request related to a problem? Please describe
There is a problem scanning real usages for non-php files. For example, Symfony uses the `container.xml` to define services. In there, there are classes linked using their FQN.

## Describe the solution you'd like
Provide an extension API so users can develop their own usage scanner to avoid false positives.

## Configuration
As this is a composer plugin, the configuration should be placed inside the `composer.json`.
We already have a key beneath the `extras` section, this can be used and needs to be adjusted.

A possible structure for the configuration could be:
```
"extras": {
  "unused": {
    "packageSymbolProviders": [
      // Fully Qualified names of custom package symbol providers implementing a certain interface.
      // These classes should be autoloaded when `vendor/autoload.php` is loaded from the root package.
      // These loaders should have a list of packages they handle for example, these packages will be ignored during the
      // default symbol scanning implementation

      "My\\Custom\\Symfony\\PackageSymbolProvider": [
        "symfony/console"
      ]
    ],
    "rootSymbolProviders": [
      // Fully Qualified names of custom root symbol providers implementing a certain interface.
      // Used to find used symbols in the root package. For example to provide used symbols in configuration files (yaml, etc)

      "My\\Custom\\YamlSymbolProvider"
    ],
    "ignorePackages": [
      // Package names to be ignored. Configuration to avoid false positives or smt else
    ],
    "skipFolders": [
      // root folders to skip from scanning (vendor is default)
    ]
  }
}
```

## Interface
The interfaces used for providing root or package symbols should be extracted into a separate composer package.
For example `composer-unused/custom-provider-interface`.

This package should contain the following interfaces:

### `RequiredPackageInterface`
Marker interface for a required package.

### `RootPackageInterface`
Marker interface for the root package `composer-unused` is checking

### `PackageSymbolProvider`
An interface for providing package symbols. This interface will receive a `RequirePackageInterface` argument with
the information about the package to provide symbols for.

### `RootSymbolProvider`
An interface for providing packages from the root package. This should be mostly used for non-php or custom configuration files where `string` is used rather than the `::class` constant.

See #74, #43

**Comment by VincentLanglet** [comment]
Hi ! 👋 
What is the news about the implementation of this issue ? Any good news ? :)

Thanks for the tool

**Comment by icanhazstring** [comment]
HI @VincentLanglet thanks for reaching out.
Unfortunately I didn't find the appropriate time to implement it.
I did however some "usecase research" on paper.

I will probably simply update this issue with some of my ideas so people can see the progress :)
And thanks again 👍

**Comment by icanhazstring** [comment]
-- Updated information about how this interface should work

**Comment by lulco** [comment]
How will I pass some parameters to providers? For example `My\\Custom\\YamlSymbolProvider` will need some list of directories to check.

Background:
I have composer-unused installed globally and use in CI for all my apps. So the only difference is the path where the app is located.
Also I have multiple types of apps - web app, library, cms module etc. Each type has different structure - in web app there is an `app` directory, in library and cms module there are `src` directories. So I would need to tell to this YamlSymbolProvider it should parse files in app or src

**Comment by icanhazstring** [comment]
That is a good question :)
The answer is, you will do it yourself.

You will add your custom provider to the configuration.
The custom parser might be located inside your code, or in another required package.
`composer-unused` will run through all providers and fetch the symbols.

The easiest way is to parse every `.yml` file from the path where your `composer.json` is located (recursively)
